### PR TITLE
Requirement's Update for FE implementation 

### PR DESCRIPTION
**Pull request by Parth858**
Addition of few endpoints are:

- [x] getuserlist 

- [x] getUsers = n no of users (count)

- [x] jobs categories = as per the posted jobs 

- [x] get posted jobs 
- [x] get specific job details with job_id..

### CODE REVIEW

File: apps/jobs/constants/values.py [hunk 1]
```diff
@@ -11,6 +11,15 @@
 
 GENDER = (("male", "Male"), ("female", "Female"), ("other", "Other"))
 
+HIRING_STATUS = (("hiring", "HIRING"), ("open to work", "OPEN TO WORK"))
+
+JOB_TYPE = (
```
Comment (hims1911) [inline apps/jobs/constants/values.py:16]: @Himan10 lets not store upper case and smaller case. lets keep it constant. either lower or upper.

File: apps/jobs/views.py [hunk 2]
```diff
@@ -279,6 +279,46 @@ def update_application(self, request, pk=None):
                 "Status has been updated!!", status.HTTP_200_OK
             )
 
+    @action(detail=False, methods=["get"])
+    def job_categories(self, request):
+        """
+        API endpoint: jobs/JobCategories
```
Comment (hims1911) [inline apps/jobs/views.py:285]: check the endpoint, are we using _ or camelcase

File: apps/jobs/views.py [hunk 3]
```diff
@@ -483,6 +523,21 @@ def get_application_status(self, serialized_data):
 
         return serialized_data
 
+    @action(detail=False, methods=["get"])
+    def get_users(self, request):
+        page = self.paginate_queryset(User.objects.all())
+        count = request.query_params.get("count", 10)  # Default count is 10
```
Comment (hims1911) [inline apps/jobs/views.py:529]: keep the constant in settings.py

Comment (hims1911) [reply apps/jobs/views.py:529]: or constants.py

**Review by hims1911 — CHANGES_REQUESTED** [review]

**Comment by Parth858** [comment]
Added new end points 
1.``get_posted_jobs`` :  specifes us the limited job data such as  [job_id, job_role, location and number of applicants] with its own independent JWT token for every job.
2.``get_job-details``: will retrieve all info related to specific job which was used with the help of JWT token.

**Review by Himan10 — CHANGES_REQUESTED** [review]

File: apps/jobs/views.py [hunk 4]
```diff
@@ -101,25 +102,14 @@ def create(self, request, *args, **kwargs):
...
-        job_data = Job.objects.filter(job_id=pk)
-        serialized_job_data = self.serializer_class(job_data, many=True)
-        if serialized_job_data:
-            serialized_job_data = self.get_number_of_applicants(serialized_job_data)
-        return response.create_response(serialized_job_data.data, status.HTTP_200_OK)
+    def retrieve(self, request, *args, **kwargs):
+        """
+        Overriding the retrieve method to redirect it to the new retrieve_job_by_id method.
+        """
+        return self.retrieve_job_by_id(request, *args, **kwargs)
```
Comment (Parth858) [inline apps/jobs/views.py:108]: This method is overridden from the base class and serves as the entry point for retrieving job data.
It redirects the request to the ``retrieve_job_by_id`` method.

File: apps/jobs/views.py [hunk 5]
```diff
@@ -141,6 +132,47 @@ def get_number_of_applicants(self, serialized_data):
...
+            jobs_belong_to_company = Job.objects.filter(
+                company_id=company.get("company_id")
+            )
+            active_jobs = sum(1 for job in jobs_belong_to_company if job.is_active)
+            company.update({"Active Jobs": active_jobs})
+
+        return serialized_company_data
+
+    @action(detail=False, methods=["get"], url_path="details")
+    def retrieve_job_by_id(self, request):
```
Comment (Parth858) [inline apps/jobs/views.py:165]: It is accessible at the endpoint ``jobs/details`` using a GET request. It retrieves the job_id from the query parameter and also validates the job_id to ensure it is a valid UUID. 

> Attempts to retrieve the job data from the database using the Job model.
If the job with the specified job_id exists, it serializes the job data and returns it in the response.
If the job does not exist, it returns a 404 Not Found response.

File: apps/jobs/views.py (hunk 4 above)
Comment (Himan10) [reply apps/jobs/views.py:108]: you don't need to create multiple methods to do the same thing, `retrieve_job_by_id` can be done within `retrieve` method as well.

File: apps/jobs/views.py [hunk 6]
```diff
@@ -156,6 +145,47 @@ def get_active_jobs_count(serialized_company_data):
...
+        try:
+            job_data = Job.objects.get(job_id=job_id)
+        except Job.DoesNotExist:
+            return response.create_response(
+                f"Job with job_id '{job_id}' does not exist", status.HTTP_404_NOT_FOUND
+            )
+
+        serialized_job_data = self.serializer_class(job_data)
+
+        return response.create_response(serialized_job_data.data, status.HTTP_200_OK)
```
Comment (Himan10) [inline apps/jobs/views.py:187]: Do not modify the existing implementation, keep it as it is. 
Here, you have excluded the key "Number of Applicants" from the `retrieve` response, which is crucial in the first place.

**Review by Himan10 — CHANGES_REQUESTED** [review]

File: apps/jobs/views.py [hunk 7]
```diff
@@ -141,7 +154,23 @@ def get_number_of_applicants(serialized_data):
             jobdata.update({"Number of Applicants": number_of_applicants})
 
         return serialized_data
-    
+
+    @staticmethod
+    def get_active_jobs_count(serialized_company_data):
```
Comment (Himan10) [inline apps/jobs/views.py:159]: was this added by you ?

File: apps/jobs/views.py [hunk 8]
```diff
@@ -141,7 +154,23 @@ def get_number_of_applicants(serialized_data):
...
+            jobs_belong_to_company = Job.objects.filter(
+                company_id=company.get("company_id")
+            )
+            active_jobs = sum(1 for job in jobs_belong_to_company if job.is_active)
+            company.update({"Active Jobs": active_jobs})
+
+        return serialized_company_data
+
     @staticmethod
     def get_active_jobs_count(serialized_company_data):
```
Comment (Himan10) [inline apps/jobs/views.py:175]: why do we have two methods with the same name ?

**Review by Himan10 — CHANGES_REQUESTED** [review]

File: apps/jobs/constants/values.py (hunk 1 above)
Comment (hims1911) [reply apps/jobs/constants/values.py:20]: you can do the check instead of keeping in variable

**Review by hims1911 — APPROVED** [review]

**Pull request merged**